### PR TITLE
I2S ADC Mode implemented for ESP32

### PR DESCRIPTION
**Pull request by MarP13**
## Description

If I2S_MODE_ADC_BUILT_IN mode is used:
- Analog channel is resolved from I2S1_MDATA_IN and correct attenuation is configured 
- I2S ADC mode is set for given channel
- ADC Mode is disable for given bus
- 

## How Has This Been Tested?
Tested on ESP32_REV3

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=2539) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=2539) before we can accept your contribution.<br/><br/>:x: MarP13 [sign now](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=2539)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nf-interpreter?pullRequest=2539) it.</sub>

**Comment by MarP13** [comment]
@BWalti @josesimoes  Hey guys, please take a look and feel free to take over. 
I tried to marked the points worth polishing by TODOs. I may eventually work on the comments you will certainly have, but factoring in mt huge lack of C++, I could easily break some of the things.
Thanks in advance
MartinP

**Comment by MProchaska-PCTY** [comment]
At this moment, I don't have much to add so I am marking the PR as ready for review to collect your feedback.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 1]
```diff
@@ -494,6 +580,11 @@ HRESULT Library_sys_dev_i2s_native_System_Device_I2s_I2sDevice::NativeDispose___
 
         if (Esp_I2S_Initialised_Flag[bus] <= 0)
         {
+#if SOC_I2S_SUPPORTS_ADC
+            // TODO call only for ADC mode? Process result?
+            i2s_adc_disable(bus);
```
Comment (BWalti) [inline targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:585]: IIRC there is a mode on the config, which could be used to decide if we need to do this. I believe you are already using `mode` to detect at some point, that ADC is used. Maybe we could re-use this here as well?

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 2]
```diff
@@ -207,6 +219,80 @@ HRESULT SetI2sConfig(i2s_port_t bus, CLR_RT_HeapBlock *config)
...
+        if (mode & I2S_MODE_ADC_BUILT_IN)
+        {
+            // TODO - make attenuation configurable?
+            adc_atten_t atten = ADC_ATTEN_DB_11;
+
+            // TODO Re-use logic in ADC?
+            int channelNumber = -1;
+
+            // Find out ADC Channel for pin confiured as I2S_DATAIN
+            for (unsigned int i = 0; i < sizeof(Esp32_ADC_DevicePinMap) / sizeof(int8_t); i++)
```
Comment (BWalti) [inline targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:234]: this extracting of the channelNumber seems to be a bit magic. I assume there would be a more straight forward way to achieve this?

Comment (BWalti) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:234]: Maybe @Ellerbach has an idea?

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 3]
```diff
@@ -207,6 +219,80 @@ HRESULT SetI2sConfig(i2s_port_t bus, CLR_RT_HeapBlock *config)
             NANOCLR_SET_AND_LEAVE(CLR_E_INVALID_PARAMETER);
         }
 
+#if SOC_I2S_SUPPORTS_ADC
+
+        // Configure ADC Mode
+        if (mode & I2S_MODE_ADC_BUILT_IN)
```
Comment (BWalti) [inline targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:225]: If we are in ADC mode, is it still "ok" to invoke `i2s_driver_install` with the config as we currently are doing?

**Comment by MarP13** [comment]
> 

@dotnet-policy-service agree

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp (hunk 2 above)
Comment (Ellerbach) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:234]: ADC only supports 1 channel: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/i2s.html#adc-dac-mode

And it's always 0

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 4]
```diff
@@ -51,6 +51,11 @@ void Esp32_I2s_UnitializeAll()
         if (Esp_I2S_Initialised_Flag[c])
         {
             // Delete bus driver
+
+#if SOC_I2S_SUPPORTS_ADC
+            // TODO call only for ADC mode? Process result?
+            i2s_adc_disable((i2s_port_t)c);
```
Comment (Ellerbach) [inline targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:57]: yes, you should call this only for ADC

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp (hunk 1 above)
Comment (Ellerbach) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:585]: you should indeed cvall the function only of the ADC mode is activated

Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp]: Done.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp (hunk 2 above)
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:239]: Is that really so? IMHO, I2S-ADC mode only supports BUS0, but it should support all channels.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 5]
```diff
@@ -207,6 +219,80 @@ HRESULT SetI2sConfig(i2s_port_t bus, CLR_RT_HeapBlock *config)
...
+                            NANOCLR_SET_AND_LEAVE(CLR_E_INVALID_OPERATION);
+                        }
+                    }
+                    break;
+
+                case 2:
+                    // Adjust for ADC2
+                    channelNumber -= 10;
+
+                    // TODO - Make ADC Resolution configurable?
```
[the remark this answers is not in the document]
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:278]: I wanted to. But BitsPerSample provides 8, 12, 16, 24 values. In contrary, ADC width has 9, 10, 11, 12. 
Nonetheless, it seems that BitsPerSample does affect downscale) the actually captured values even without calling that method.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp (hunk 4 above)
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp]: done

**Comment by Ellerbach** [comment]
@MarP13  the build is failing for some of the targets most likely because they do not have support for it. See other features like touch pin for example where there are cases depending on the device.

**Comment by MarP13** [comment]
@Ellerbach @BWalti 
I have addressed unsupported chipsets and hopefully other comments too. 
Let me please known if there is anything else to cover.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp (hunk 2 above)
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:248]: I have added the validation for BUS1 in the previous commits. So I consider this complete.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:287]: My suggestion here would be to have this as potential improvement later.

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 6]
```diff
@@ -30,6 +30,8 @@ static char Esp_I2S_Initialised_Flag[I2S_NUM_MAX] = {
 #endif
 };
 
+static char Adc_Mode_Enabled = 0;
```
[the remark this answers is not in the document]
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp]: Done together with few minor improvements.

**Review by Ellerbach — APPROVED** [review]
Thanks a lot, it looks good to me! @josesimoes do you want to have a quick look?

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 7]
```diff
@@ -207,10 +224,90 @@ HRESULT SetI2sConfig(i2s_port_t bus, CLR_RT_HeapBlock *config)
             NANOCLR_SET_AND_LEAVE(CLR_E_INVALID_PARAMETER);
         }
 
+#if SOC_I2S_SUPPORTS_ADC
```
[the remark this answers is not in the document]
Comment (MarP13) [reply targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:236]: Added here https://github.com/nanoframework/nf-interpreter/pull/2539/commits/6ccc12e17534713f078eaf12084bf98185c8c21f

**Comment by MarP13** [comment]
> Thanks a lot, it looks good to me! @josesimoes do you want to have a quick look?

@josesimoes Please let me know if there is more to cover. Thx.

**Review by josesimoes — APPROVED** [review]
LGTM! Thank you for making this available. 👍🏻

File: targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp [hunk 8]
```diff
@@ -207,10 +233,90 @@ HRESULT SetI2sConfig(i2s_port_t bus, CLR_RT_HeapBlock *config)
...
+        // Configure ADC Mode
+        if (mode & I2S_MODE_ADC_BUILT_IN)
+        {
+            // TODO - make attenuation configurable?
+            adc_atten_t atten = ADC_ATTEN_DB_11;
+
+            // TODO Re-use logic in ADC?
+            int channelNumber = -1;
+
+            // Find out ADC Channel for pin confiured as I2S_DATAIN
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/System.Device.I2s/sys_dev_i2s_native_System_Device_I2s_I2sDevice.cpp:247]:
```suggestion
            // Find out ADC Channel for pin configured as I2S_DATAIN
```

**Pull request merged**